<!DOCTYPE html>
<html lang="en-us">

  <head>
  <link href="http://gmpg.org/xfn/11" rel="profile">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta http-equiv="content-type" content="text/html; charset=utf-8">

  <!-- Enable responsiveness on mobile devices-->
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1">

  <title>
    
      Call variants &middot; Training Course on Next generation sequencing
    
  </title>

  <!-- CSS -->
    <!-- Latest compiled and minified CSS -->
  <link rel="stylesheet" href="/public/css/poole.css">
  <link rel="stylesheet" href="/public/css/syntax.css">
  <link rel="stylesheet" href="/public/css/lanyon.css">
  <link rel="stylesheet" href="/public/css/styles.css">
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=PT+Serif:400,400italic,700%7CPT+Sans:400">

  <!-- Icons -->
  <link rel="apple-touch-icon-precomposed" sizes="144x144" href="/public/apple-touch-icon-precomposed.png">
  <link rel="shortcut icon" href="/public/favicon.ico">

  <!-- RSS -->
  <link rel="alternate" type="application/rss+xml" title="RSS" href="/atom.xml">

  <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.1.4/jquery.min.js"></script>
  <script src="/js/toc.js"></script>
</head>


  <body class="theme-base-07">

    <!-- Target for toggling the sidebar `.sidebar-checkbox` is for regular
     styles, `#sidebar-checkbox` for behavior. -->
<input type="checkbox" class="sidebar-checkbox" id="sidebar-checkbox" checked>

<!-- Toggleable sidebar -->
<div class="sidebar" id="sidebar">
  <div class="sidebar-item">
    <p>NGS training</p>
  </div>

  <nav class="sidebar-nav">
    <a class="sidebar-nav-item" href="/">Home</a>

    

    
    
      
        
      
    
      
    
      
        
          <a class="sidebar-nav-item" href="http://localhost:4000/lectures/01.welcome/">Welcome and introduction</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="http://localhost:4000/lectures/02.setup_tutorial/">Enviroment installation</a>
        
      
    
      
    
      
        
          <a class="sidebar-nav-item" href="http://localhost:4000/lectures/05.quality_control/">Quality control & Adapter trimming</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="http://localhost:4000/lectures/06.bioinformatics_pipeline/">Bioinformatics approach examples</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="http://localhost:4000/lectures/07.rnaseq_pipeline/">RNA-Seq data analysis</a>
        
      
    
      
        
          <a class="sidebar-nav-item active" href="http://localhost:4000/lectures/08.callvariant/">Call variants</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="http://localhost:4000/lectures/10.docker/">Everything in a box. Docker and Galaxy</a>
        
      
    

    <!--<a class="sidebar-nav-item" href="/archive/v2.0.zip">Download</a>-->
    <a class="sidebar-nav-item" href="https://bioinfo-dirty-jobs.github.io/NGSBASE">GitHub project</a>
    <span class="sidebar-nav-item">Currently v2.0</span>
  </nav>

  <div class="sidebar-item">
    <p>
      &copy; 2019. All rights reserved.
    </p>
  </div>
</div>


    <!-- Wrap is the content to shift when toggling the sidebar. We wrap the
         content to avoid any CSS collisions with our real content. -->
    <div class="wrap">
      <div class="masthead">
        <div class="container">
          <h3 class="masthead-title">
            <a href="http://localhost:4000/" title="Home">Training Course on Next generation sequencing</a>
            <!--<p></p><small>Call variants</small></p>-->
          </h3>
        </div>
      </div>

      <div class="container content">
        <div class="page">
  <h1 class="page-title">Call variants</h1>
  <p class="note">Call variants.</p>
  
    <!-- this handles the automatic toc. use ## for subheads to auto-generate the on-page minitoc. if you use html tags, you must supply an ID for the heading element in order for it to appear in the minitoc. -->
<script>
$( document ).ready(function() {
  // Handler for .ready() called.

$('#toc').toc({ minimumHeaders: 0, listType: 'ul', showSpeed: 0, headers: 'h2,h3,h4' });

/* this offset helps account for the space taken up by the floating toolbar. */
$('#toc').on('click', 'a', function() {
  var target = $(this.getAttribute('href'))
    , scroll_target = target.offset().top

  $(window).scrollTop(scroll_target - 10);
  return false
})

});
</script>

<div id="toc"></div>

  
  <h2 id="sofwtare">Sofwtare</h2>

<h2 id="software-used-in-this-practical">Software used in this practical:</h2>
<hr />

<ul>
  <li><a href="http://samtools.sourceforge.net/" title="samtools">SAMTools</a> : SAM Tools provide various utilities for manipulating alignments in the SAM format, including sorting, merging, indexing and generating alignments in a per-position format.</li>
  <li><a href="http://picard.sourceforge.net/" title="Picard">Picard</a> : Picard comprises Java-based command-line utilities that manipulate SAM files, and a Java API (SAM-JDK) for creating new programs that read and write SAM files.</li>
  <li><a href="http://www.broadinstitute.org/gatk/" title="GATK">GATK</a> : Genome Analysis Toolkit - A package to analyse next-generation re-sequencing data, primary focused on variant discovery and genotyping.</li>
</ul>

<h2 id="file-formats-explored">File formats explored:</h2>

<ul>
  <li><a href="http://samtools.sourceforge.net/SAMv1.pdf">SAM</a></li>
  <li><a href="http://www.broadinstitute.org/igv/bam">BAM</a></li>
  <li>VCF Variant Call Format: see <a href="http://www.1000genomes.org/wiki/analysis/variant-call-format/vcf-variant-call-format-version-42">1000 Genomes</a> and <a href="http://en.wikipedia.org/wiki/Variant_Call_Format">Wikipedia</a> specifications.</li>
</ul>

<p><img src="http://localhost:4000/images/gatkgermline.png" /></p>

<h1 id="exercise-2-variant-calling-with-pair-end-data">Exercise 2: Variant calling with pair-end data</h1>

<p>Go to the example 1 folder in your course directory:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>cd ~/Desktop/Data/CALL_VARIANT/
</code></pre></div></div>

<h1 id="prepare-bam-file">Prepare BAM file</h1>

<p>We must sort the BAM file using <code class="highlighter-rouge">samtools</code>:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>samtools sort example1/000-dna_chr21_100_hq_se.bam  -o example1/001-dna_chr21_100_hq_se_sorted.bam
</code></pre></div></div>

<p>Index the BAM file:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>samtools index example1/001-dna_chr21_100_hq_se_sorted.bam
</code></pre></div></div>

<h1 id="mark-duplicates-using-picard">Mark duplicates (using Picard)</h1>
<p>Mark and remove duplicates:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>picard MarkDuplicates \
		INPUT=example1/001-dna_chr21_100_hq_se_sorted.bam \
		OUTPUT=example1/001-dna_chr21_100_hq_se_sorted_noDup.bam \
		METRICS_FILE=002-metrics.txt
</code></pre></div></div>

<p>Index the new BAM file:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>picard BuildBamIndex INPUT=example1/001-dna_chr21_100_hq_se_sorted_noDup.bam 
</code></pre></div></div>

<h1 id="local-realignment-around-indels-using-gatk">Local realignment around INDELS (using GATK)</h1>

<p>There are 2 steps to the realignment process:</p>

<p>Create a target list of intervals which need to be realigned</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>java -jar /home/bioinfo2/Data/Soft/GenomeAnalysisTK.jar \
	-T RealignerTargetCreator \
	-R genome/f000_chr21_ref_genome_sequence.fa \
	-Iexample1/001-dna_chr21_100_hq_se_sorted_noDup.bam\
	-o 003-indelRealigner.intervals
</code></pre></div></div>

<p>Perform realignment of the target intervals:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>java -jar /home/bioinfo2/Data/Soft/GATK/GenomeAnalysisTK.jar \
	-T IndelRealigner \
	-R genome/f000_chr21_ref_genome_sequence.fa \
	-I example1/001-dna_chr21_100_hq_se_sorted_noDup.bam \
	-targetIntervals 003-indelRealigner.intervals \
	-o 003-dna_chr21_100_hq_se_sorted_noDup_realigned.bam
</code></pre></div></div>

<h1 id="base-quality-score-recalibration-using-gatk">Base quality score recalibration (using GATK)</h1>
<hr />

<p>Two steps:</p>

<p>Analyse patterns of covariation in the sequence dataset</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>java -jar /home/bioinfo2/Data/Soft/GenomeAnalysisTK.jar \
	-T BaseRecalibrator \
	-R  genome/f000_chr21_ref_genome_sequence.fa \
	-I 003-dna_chr21_100_hq_se_sorted_noDup_realigned.bam \
	-knownSites 000-dbSNP_chr21.vcf \
	-o 004-recalibration_data.table
</code></pre></div></div>

<p>Apply the recalibration to your sequence data</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>java -jar /home/bioinfo2/Data/Soft/GenomeAnalysisTK.jar \
	-T PrintReads \
	-R genome/f000_chr21_ref_genome_sequence.fa  \
	-I 003-dna_chr21_100_hq_se_sorted_noDup_realigned.bam \
	-BQSR 004-recalibration_data.table \
	-o 004-dna_chr21_100_hq_se_sorted_noDup_realigned_recalibrated.bam
</code></pre></div></div>

<h1 id="variant-calling-using-gatk---haplotypecaller">Variant calling (using GATK - <strong>HaplotypeCaller</strong>)</h1>
<hr />

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>java -jar /home/bioinfo2/Data/Soft/GenomeAnalysisTK.jar \
	-T HaplotypeCaller \
	-R genome/f000_chr21_ref_genome_sequence.fa \
	-I 004-dna_chr21_100_hq_se_sorted_noDup_realigned_recalibrated.bam \
	-o 005-dna_chr21_100_he_se.vcf
</code></pre></div></div>

<p>Once we have called the variants, we might be interested in filtering out some according to our confidence in them. Some of the most basic filters consist of identifying variants with low calling quality or a low number of reads supporting the variant. There are many programs that can be used to filter VCFs. Here we are going to use bcftools from Samtools to preform a basic filtering.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>bcftools filter -s LowQual -e 'QUAL&lt;20 | DP&lt;3' 005-dna_chr21_100_he_se.vcf &gt; 005-dna_chr21_100_he_se_filtered.vcf
</code></pre></div></div>

<p>Let’s see how many variants fail to pass our filters:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code> grep LowQual 005-dna_chr21_100_he_se_filtered.vcf | wc -l
</code></pre></div></div>

<p>And how many passed:</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>  grep PASS 005-dna_chr21_100_he_se_filtered.vcf| wc -l
</code></pre></div></div>

<h1 id="somatic-call">SOMATIC CALL</h1>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>cd  somatic_calling
samtools sort 000-normal.bam -o 001-normal_sorted.bam
samtools sort 000-tumor.bam -o  001-tumor_sorted.bam
samtools index 001-normal_sorted.bam
samtools index 001-tumor_sorted.bam
java -jar ~/soft/GATK/GenomeAnalysisTK.jar \
--analysis_type MuTect \
--reference_sequence TP53.hg19.fa \
--dbsnp 000-dbsnp_132_b37.leftAligned.vcf.gz \
--cosmic 000-b37_cosmic_v54_120711.vcf.gz \
--input_file:normal 001-normal_sorted.bam \
--input_file:tumor 001-tumor_sorted.bam \
--out 002-call_stats.out \
--coverage_file 002-coverage.wig \
--vcf 003-somatic_variants.vcf
</code></pre></div></div>

</div>

      </div>
    </div>

    <label for="sidebar-checkbox" class="sidebar-toggle"></label>

    <script>
      (function(document) {
        var toggle = document.querySelector('.sidebar-toggle');
        var sidebar = document.querySelector('#sidebar');
        var checkbox = document.querySelector('#sidebar-checkbox');

        document.addEventListener('click', function(e) {
          var target = e.target;

          if(!checkbox.checked ||
             sidebar.contains(target) ||
             (target === checkbox || target === toggle)) return;

          checkbox.checked = false;
        }, false);
      })(document);
    </script>
  <!--<footer>
        <p>This project is maintained by  Maurizio Polano<br/></p>
        <p><small>Hosted on GitHub Pages &mdash; Theme by <a href="https://github.com/orderedlist">orderedlist</a></small></p>
      </footer>
    </div>
    <script src="javascripts/scale.fix.js"></script>

-->
  </body>
</html>
