<!DOCTYPE html>
<html lang="en-us">

  <head>
  <link href="http://gmpg.org/xfn/11" rel="profile">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta http-equiv="content-type" content="text/html; charset=utf-8">

  <!-- Enable responsiveness on mobile devices-->
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1">

  <title>
    
      Everything in a box. Docker and Galaxy &middot; Training Course on Next generation sequencing
    
  </title>

  <!-- CSS -->
    <!-- Latest compiled and minified CSS -->
  <link rel="stylesheet" href="/public/css/poole.css">
  <link rel="stylesheet" href="/public/css/syntax.css">
  <link rel="stylesheet" href="/public/css/lanyon.css">
  <link rel="stylesheet" href="/public/css/styles.css">
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=PT+Serif:400,400italic,700%7CPT+Sans:400">

  <!-- Icons -->
  <link rel="apple-touch-icon-precomposed" sizes="144x144" href="/public/apple-touch-icon-precomposed.png">
  <link rel="shortcut icon" href="/public/favicon.ico">

  <!-- RSS -->
  <link rel="alternate" type="application/rss+xml" title="RSS" href="/atom.xml">

  <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.1.4/jquery.min.js"></script>
  <script src="/js/toc.js"></script>
</head>


  <body class="theme-base-07">

    <!-- Target for toggling the sidebar `.sidebar-checkbox` is for regular
     styles, `#sidebar-checkbox` for behavior. -->
<input type="checkbox" class="sidebar-checkbox" id="sidebar-checkbox" checked>

<!-- Toggleable sidebar -->
<div class="sidebar" id="sidebar">
  <div class="sidebar-item">
    <p>NGS training</p>
  </div>

  <nav class="sidebar-nav">
    <a class="sidebar-nav-item" href="/">Home</a>

    

    
    
      
        
      
    
      
    
      
        
          <a class="sidebar-nav-item" href="http://localhost:4000/lectures/01.welcome/">Welcome and introduction</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="http://localhost:4000/lectures/02.setup_tutorial/">Enviroment installation</a>
        
      
    
      
    
      
        
          <a class="sidebar-nav-item" href="http://localhost:4000/lectures/05.quality_control/">Quality control & Adapter trimming</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="http://localhost:4000/lectures/06.bioinformatics_pipeline/">Bioinformatics approach examples</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="http://localhost:4000/lectures/07.rnaseq_pipeline/">RNA-Seq data analysis</a>
        
      
    
      
        
          <a class="sidebar-nav-item" href="http://localhost:4000/lectures/08.callvariant/">Call variants</a>
        
      
    
      
        
          <a class="sidebar-nav-item active" href="http://localhost:4000/lectures/10.docker/">Everything in a box. Docker and Galaxy</a>
        
      
    

    <!--<a class="sidebar-nav-item" href="/archive/v2.0.zip">Download</a>-->
    <a class="sidebar-nav-item" href="https://bioinfo-dirty-jobs.github.io/NGSBASE">GitHub project</a>
    <span class="sidebar-nav-item">Currently v2.0</span>
  </nav>

  <div class="sidebar-item">
    <p>
      &copy; 2019. All rights reserved.
    </p>
  </div>
</div>


    <!-- Wrap is the content to shift when toggling the sidebar. We wrap the
         content to avoid any CSS collisions with our real content. -->
    <div class="wrap">
      <div class="masthead">
        <div class="container">
          <h3 class="masthead-title">
            <a href="http://localhost:4000/" title="Home">Training Course on Next generation sequencing</a>
            <!--<p></p><small>Everything in a box. Docker and Galaxy</small></p>-->
          </h3>
        </div>
      </div>

      <div class="container content">
        <div class="page">
  <h1 class="page-title">Everything in a box. Docker and Galaxy</h1>
  <p class="note">Docker is an open platform for developing, shipping, and running applications. Galaxy is available as Docker Image, an easy distributable full-fledged Galaxy installation. Finally, Galaxy supports running tools within Docker containers.</p>
  
    <!-- this handles the automatic toc. use ## for subheads to auto-generate the on-page minitoc. if you use html tags, you must supply an ID for the heading element in order for it to appear in the minitoc. -->
<script>
$( document ).ready(function() {
  // Handler for .ready() called.

$('#toc').toc({ minimumHeaders: 0, listType: 'ul', showSpeed: 0, headers: 'h2,h3,h4' });

/* this offset helps account for the space taken up by the floating toolbar. */
$('#toc').on('click', 'a', function() {
  var target = $(this.getAttribute('href'))
    , scroll_target = target.offset().top

  $(window).scrollTop(scroll_target - 10);
  return false
})

});
</script>

<div id="toc"></div>

  
  <p><img src="http://localhost:4000/images/06_docker_logo.png" alt="Docker_logo" /></p>

<h2 id="introduction">Introduction</h2>

<p>Linux Containers (LXC) is an operating-system-level virtualization method for running multiple isolated Linux systems (named containers) on a single control host (LXC host). It does not provide a virtual machine, but rather provides a virtual environment that has its own CPU, memory, block I/O, network, etc. space and the resource control mechanism. This is provided by namespaces and cgroups features in Linux kernel on LXC host.</p>

<p>Docker is an open-source engine to easily create lightweight, portable, self-sufficient containers from any application. The same container that a developer builds and test on a laptop can run at scale, in production, on VMs, OpenStack cluster, public clouds and more.</p>

<p>Docker can run in a VM (or not).</p>

<p>Containers and VMs are similar in their goals: to isolate an application and its dependencies into a self-contained unit that can run anywhere. The main difference between containers and VMs is in their architectural approach:</p>

<p><img src="http://localhost:4000/images/06_container_vs_vm.png" alt="Virtual machines vs Containers schema" /></p>

<p>Docker features:</p>
<ul>
  <li>Less overheads : isolation vs virtualization</li>
  <li>Versioning (git-like) + common base images</li>
  <li>Component re-use</li>
  <li>Nice command line based UX</li>
  <li>Simple reproducible builds and deployment</li>
  <li>Sharing (public repository)</li>
</ul>

<p>Images and Layers:</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ubuntu : 200 Mb
ubuntu + R : 250 Mb
ubuntu + matlab : 250 Mb

All three: 300 Mb
</code></pre></div></div>

<h2 id="docker-getting-started">Docker: getting started</h2>

<h3 id="installation">Installation</h3>

<p>Docker is already installed on your VMs.
To install it on your systems:</p>
<ol>
  <li><a href="https://docs.docker.com/engine/installation/linux/ubuntu/">Get Docker on Ubuntu</a></li>
  <li><a href="https://docs.docker.com/engine/installation/linux/linux-postinstall/">Post-installation steps for Linux</a></li>
</ol>

<h3 id="search-pull-run-and-more">Search, pull, run and more</h3>

<p><img src="http://localhost:4000/images/06_docker_command.png" alt="Docker_client" /></p>

<p>Examples:</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ docker search debian
ubuntu@workspace-new:~$ docker search debian
NAME                                         DESCRIPTION                                     STARS     OFFICIAL   AUTOMATED
debian                                       Debian is a Linux distribution that's comp...   2129      [OK]


$ docker pull debian:jessie
jessie: Pulling from library/debian
9f0706ba7422: Pull complete 
Digest: sha256:4bc62f74d246e8428be8dd3833461ba2cfd135064aed4001f3c12b87a011e30c
Status: Downloaded newer image for debian:jessie


$ docker images
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
debian              jessie              62a932a5c143        4 days ago          123 MB
</code></pre></div></div>

<h2 id="running-containers">Running containers</h2>

<p>Docker runs processes in isolated containers. A container is a process which runs on a host. The host may be local or remote. When an operator executes docker run, the container process that runs is isolated in that it has its own file system, its own networking, and its own isolated process tree separate from the host.
The basic docker run command:</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ docker run [OPTIONS] IMAGE[:TAG|@DIGEST] [COMMAND] [ARG...]
</code></pre></div></div>
<p>Example:</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ docker run --rm python:3.5 python -c "print(40 + 2)"

Unable to find image 'python:3.5' locally
latest: Pulling from library/python


357ea8c3d80b: Already exists
52befadefd24: Pull complete
3c0732d5313c: Pull complete
ceb711c7e301: Pull complete
4211bb537697: Pull complete
71f9074c0739: Pull complete
3e5349707036: Pull complete
Digest: sha256:a755ad5a30b2[...]
Status: Downloaded newer image for python:3.5
42

$ docker run --rm python:3.5 python -c "print(40 + 3)"
43
</code></pre></div></div>

<p><a href="https://docs.docker.com/engine/reference/run/#clean-up-rm">–rm option</a></p>

<h3 id="run-an-interactive-container">Run an interactive container</h3>

<p>In foreground mode (the default when -d is not specified), docker run can start the process in the container and attach the console to the process’s standard input, output, and standard error. It can even pretend to be a TTY, this is what most command line executables expect</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ docker run --rm -ti python:3.5 bash
root@10d2dfedb935:/# ps
PID TTY   TIME  CMD
1 ?  00:00:00 bash
8 ?  00:00:00 ps


root@10d2dfedb935:/# python
Python 3.5.2 (default, Aug  9 2016, 20:58:38)
[GCC 4.9.2] on linux
&gt;&gt;&gt; 40 + 2
42
&gt;&gt;&gt;
root@10d2dfedb935:/# exit
</code></pre></div></div>

<h3 id="run-in-detached-mode">Run in detached mode</h3>

<p>To start a container in detached mode (background), use -d option. By design, containers started in detached mode exit when the root process used to run the container exits.</p>

<p>Example: publish nginx container port 80 on 8080 port</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ docker run -d -p 8080:80 nginx
Unable to find image 'nginx:latest' locally
latest: Pulling from library/nginx
e6e142a99202: Pull complete 
8c317a037432: Pull complete 
af2ddac66ed0: Pull complete 
Digest: sha256:72c7191585e9b79cde433c89955547685db00f3a8595a750339549f6acef7702
Status: Downloaded newer image for nginx:latest                                                                                                                                                                                  
cc1412bff7ebf42f4173a81ee744c567b24079708ce701494faeabd645866a45

$ docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS                  NAMES                                                                                            
cc1412bff7eb        nginx               "nginx -g 'daemon ..."   3 minutes ago       Up 3 minutes        0.0.0.0:8080-&gt;80/tcp   ecstatic_hypatia

$ docker exec -it ecstatic_hypatia /bin/bash                                                                                                                                                               
root@cc1412bff7eb:/# nginx -t
nginx: the configuration file /etc/nginx/nginx.conf syntax is ok                                                                                                                                                                 
nginx: configuration file /etc/nginx/nginx.conf test is successful                                                                                                                                                               
root@cc1412bff7eb:/#
</code></pre></div></div>

<h3 id="inspect-containers">Inspect containers</h3>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ docker ps   #shows running containers
$ docker inspect   #info on a container (incl. IP address)
$ docker logs   #gets logs from container
$ docker events   #gets events from container
$ docker port   #shows public facing port of container
$ docker top   #shows running processes in container
$ docker diff   #shows changed files in container's FS
$ docker stats   #shows metrics, memory, cpu, filsystem
</code></pre></div></div>

<h3 id="hands-on-1">Hands On! (1)</h3>

<p>Play with biocontainers/samtools:1.3.1:</p>

<ol>
  <li>Get the image</li>
  <li>Launch samtools container interactively
    <ul>
      <li>Print the help page for samtools</li>
    </ul>

    <div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ docker run -t -i biocontainers/samtools:1.3.1 /bin/bash

$ biodocker@4a70f09adce2:/data$ samtools --help
</code></pre></div>    </div>
  </li>
  <li>Launch a samtools container in detach mode
    <ul>
      <li>Check if it exists and find its name</li>
      <li>Stop it and restart it</li>
      <li>Print the help page using this container</li>
    </ul>

    <div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ docker run biocontainers/samtools:1.3.1 samtools --help
</code></pre></div>    </div>
  </li>
</ol>

<h3 id="data-management">Data management</h3>

<p><img src="http://localhost:4000/images/06_docker_run_volume.png" alt="Docker_data_volume" /></p>

<h3 id="hands-on-2">Hands On! (2)</h3>

<p>Play with biocontainers/samtools:1.3.1:</p>

<ol>
  <li>Create locally a samtool_dir directory</li>
  <li>Launch an interactive container with a volume pointing at samtool_dir directory
    <ul>
      <li>Add a <a href="https://raw.githubusercontent.com/samtools/samtools/develop/examples/toy.sam">toy sam file</a> into the local samtool_dir directory</li>
      <li>Check if the existence of the file from inside the container</li>
      <li>Visualize the content of the toy file with samtools view commands</li>
      <li>Generate stats of the toy file into a toy_stat file into the local samtool_dir directory</li>
    </ul>
  </li>
</ol>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ mkdir -p samtools_dir
$ cd samtools_dir
$ wget https://raw.githubusercontent.com/samtools/samtools/develop/examples/toy.sam

$ docker run -it -v $HOME/samtools_dir:/data biocontainers/samtools:1.3.1 /bin/bash

biodocker@cb5912bcd1be:/data$ samtools view toy.sam 
r001    163     ref     7       30      8M4I4M1D3M      =       37      39      TTAGATAAAGAGGATACTG     *       XX:B:S,12561,2,20,112
r002    0       ref     9       30      1S2I6M1P1I1P1I4M2I      *       0       0       AAAAGATAAGGGATAAA       *
r003    0       ref     9       30      5H6M    *       0       0       AGCTAA  *
r004    0       ref     16      30      6M14N1I5M       *       0       0       ATAGCTCTCAGC    *
r003    16      ref     29      30      6H5M    *       0       0       TAGGC   *
r001    83      ref     37      30      9M      =       7       -39     CAGCGCCAT       *
x1      0       ref2    1       30      20M     *       0       0       AGGTTTTATAAAACAAATAA    ????????????????????
x2      0       ref2    2       30      21M     *       0       0       GGTTTTATAAAACAAATAATT   ?????????????????????
x3      0       ref2    6       30      9M4I13M *       0       0       TTATAAAACAAATAATTAAGTCTACA      ??????????????????????????
x4      0       ref2    10      30      25M     *       0       0       CAAATAATTAAGTCTACAGAGCAAC       ?????????????????????????
x5      0       ref2    12      30      24M     *       0       0       AATAATTAAGTCTACAGAGCAACT        ????????????????????????
x6      0       ref2    14      30      23M     *       0       0       TAATTAAGTCTACAGAGCAACTA ???????????????????????

biodocker@cb5912bcd1be:/data$ samtools flagstat toy.sam
12 + 0 in total (QC-passed reads + QC-failed reads)
0 + 0 secondary
0 + 0 supplementary
0 + 0 duplicates
12 + 0 mapped (100.00% : N/A)
2 + 0 paired in sequencing
1 + 0 read1
1 + 0 read2
2 + 0 properly paired (100.00% : N/A)
2 + 0 with itself and mate mapped
0 + 0 singletons (0.00% : N/A)
0 + 0 with mate mapped to a different chr
0 + 0 with mate mapped to a different chr (mapQ&gt;=5)


biodocker@cb5912bcd1be:/data$ samtools stats toy.sam &gt; toy_stat

biodocker@cb5912bcd1be:/data$ exit

~/samtools_dir$ ls
toy.sam  toy_stat

</code></pre></div></div>

<h3 id="creation-of-a-new-image-dockerfile">Creation of a new image: Dockerfile</h3>

<p>A Dockerfile is a script, composed of various commands (instructions) and arguments listed successively to automatically perform actions on a base image in order to create (or form) a new one.</p>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>  FROM ubuntu

  MAINTAINER Romin Irani (email@domain.com)

  RUN apt-get update

  RUN apt-get install -y nginx

  ENTRYPOINT [“/usr/sbin/nginx”,”-g”,”daemon off;”] # An ENTRYPOINT allows you to configure a container that will run as an executable.

  EXPOSE 80
</code></pre></div></div>

<p>Build docker image:</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>  $ docker build -t my_nginx_image --no-cache .
</code></pre></div></div>

<p>Login on Docker Hub with your credentials:</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>  $ docker login                                                                                                                                                                                             
    Login with your Docker ID to push and pull images from Docker Hub. If you don't have a Docker ID, head over to https://hub.docker.com to create one.                                                                             
    Username: mtangaro                                                                                                                                                                                                               
    Password:                                                                                                                                                                                                                        
    Login Succeeded  
</code></pre></div></div>

<p>Ship a Docker Image:</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>  docker push repository_name/my_nginx_image
</code></pre></div></div>

<h3 id="docker-hub-automated-build">Docker Hub Automated build</h3>

<p>Prerequisites:
To perform Docker Hub automatic build you need a Docker Hub and on the hosted repository provider (GitHub or Bitbucket) accounts.</p>

<ol>
  <li>
    <p>Create new Github (or bitbucket) repository.</p>
  </li>
  <li>
    <p>Upload the Dockerfile on GitHub/Bitbucket.
<img src="http://localhost:4000/images/06_GitHub.png" alt="Github" />
<img src="http://localhost:4000/images/06_GitHub_Dockerfile.png" alt="GitHub_Dockerfile" /></p>
  </li>
  <li>
    <p>Login on Docker Hub and select “Create Automated Build” from Create menu.
<img src="http://localhost:4000/images/06_Create_automated_build.png" alt="Create_automated_build" /></p>
  </li>
  <li>
    <p>Link Github/Bitbucket repository to Docker Hub.</p>
  </li>
  <li>
    <p>Select your repository with the Dockerfile.
<img src="http://localhost:4000/images/06_docker_hub_select_repo.png" alt="select_repo" /></p>
  </li>
  <li>
    <p>Trigger a build from “Build Settings” tab.
<img src="http://localhost:4000/images/06_trigger_build.png" alt="trigger_build" /></p>
  </li>
  <li>
    <p>Pull or Run your image.
<img src="http://localhost:4000/images/06_image_ready.png" alt="image_ready" /></p>
  </li>
</ol>

<h2 id="galaxy-docker-container">Galaxy Docker container</h2>

<p>A Docker launching a Galaxy instance and</p>

<ul>
  <li>FTP-Server</li>
  <li>Webserver</li>
  <li>Scheduler</li>
  <li>Process control UI</li>
  <li>ToolShed ready</li>
  <li>Interactive Environment ready</li>
</ul>

<p><a href="http://bgruening.github.io/docker-galaxy-stable/">Link to galaxy docker stable</a></p>

<p><a href="http://bgruening.github.io/docker-galaxy-stable/usage.html">Link to galaxy docker stabke usage</a></p>

<p>Run Galaxy Docker:</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ docker run -d -p 9080:80 -p 9021:21 --name galaxy-stable bgruening/galaxy-stable

$ docker ps
CONTAINER ID        IMAGE                     COMMAND              CREATED             STATUS              PORTS                                                                     NAMES
0e066ba9b720        bgruening/galaxy-stable   "/usr/bin/startup"   12 seconds ago      Up 11 seconds       443/tcp, 8800/tcp, 9002/tcp, 0.0.0.0:9021-&gt;21/tcp, 0.0.0.0:9080-&gt;80/tcp   galaxy-stable

$ docker exec -it galaxy-stable /bin/bash
root@0e066ba9b720:/galaxy-central#
</code></pre></div></div>

<p><img src="http://localhost:4000/images/06_docker_galaxy.png" alt="docker_galaxy" /></p>

<h3 id="hands-on-3">Hands On! (3)</h3>

<p>Launch a Galaxy Docker Container and try to:</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ docker run -d -p 9080:80 -p 9021:21 --name galaxy-stable bgruening/galaxy-stable
</code></pre></div></div>

<ul>
  <li>
    <p>Add Data</p>
  </li>
  <li>
    <p>Register and <a href="http://bgruening.github.io/docker-galaxy-stable/users-passwords.html">Become an Admin</a></p>
  </li>
  <li>Restart Galaxy
    <ul>
      <li>Hint: edit galaxy.ini file then restart docker with
        <div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>docker restart [OPTIONS] CONTAINER [CONTAINER...]
</code></pre></div>        </div>
      </li>
    </ul>
  </li>
  <li>Use export mounts (persistent data)
    <div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>docker run -d -p 9080:80 -p 9021:21 -v $HOME/galaxy_storage/:/export/ bgruening/galaxy-stable
</code></pre></div>    </div>
  </li>
  <li>Trouble shooting (logging)
    <div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>docker run -d -p 9080:80 -p 9021:21 -e "GALAXY_LOGGING=full" bgruening/galaxy-stable

docker exec -it &lt;container name&gt; bash
</code></pre></div>    </div>
    <p>Once connected to the container, log files are available in /home/galaxy/logs.</p>
  </li>
  <li>Install tools
…</li>
</ul>

<h3 id="exteding-the-docker-image-galaxy-flavors">Exteding the Docker image: Galaxy flavors</h3>

<p>Tools that are already included in the Tool Shed, can be installed to customize Galaxy docker image (Galaxy flavor) with the following steps.</p>

<h3 id="list-of-galaxy-flavors">List of Galaxy flavors</h3>

<p><a href="https://github.com/bgruening/docker-galaxy-blast">NCBI-Blast</a><br />
<a href="https://github.com/bgruening/docker-recipes/blob/master/galaxy-chemicaltoolbox">ChemicalToolBox</a><br />
<a href="https://github.com/anhi/docker-scripts/tree/master/ballaxy">ballaxy</a><br />
<a href="https://github.com/bgruening/docker-recipes/blob/master/galaxy-deeptools">NGS-deepTools</a><br />
<a href="https://github.com/gregvonkuster/docker-galaxy-ChIP-exo">Galaxy ChIP-exo</a><br />
<a href="https://github.com/bgruening/docker-galaxyp">Galaxy Proteomics</a><br />
<a href="https://github.com/bgruening/docker-galaxy-imaging">Imaging</a><br />
<a href="https://github.com/gregvonkuster/docker-galaxy-csg">Constructive Solid Geometry</a><br />
<a href="https://github.com/bgruening/galaxy-metagenomics">Galaxy for metagenomics</a><br />
<a href="https://github.com/lappsgrid-incubator/docker-galaxy-lappsgrid">Galaxy with the Language Application Grid tools</a><br />
<a href="https://github.com/gianlucacorrado/galaxy-RNAcommender">RNAcommender</a><br />
<a href="https://github.com/bgruening/galaxy-open-molecule-generator">OpenMoleculeGenerator</a><br />
<a href="https://github.com/workflow4metabolomics/w4m-docker">Workflow4Metabolomics</a><br />
<a href="https://github.com/maxplanck-ie/docker-galaxy-hicexplorer">HiC-Explorer</a><br />
<a href="https://github.com/phac-nml/snvphyl-galaxy">SNVPhyl</a><br />
<a href="https://github.com/BackofenLab/docker-galaxy-graphclust">GraphClust</a><br />
<a href="https://github.com/bgruening/galaxy-rna-workbench">RNA workbench</a><br />
<a href="https://github.com/morinlab/tools-morinlab/tree/master/docker">Cancer Genomics Toolkit</a></p>

<h2 id="integrating-docker-based-tools-within-galaxy">Integrating Docker-based tools within Galaxy</h2>

<p>Docker is a method for wrapping up a tool along with all of it’s dependencies into a single container which can be distributed with the help of Docker-Hub. A method to run docker based tools was added to Galaxy with <a href="https://bitbucket.org/galaxy/galaxy-central/pull-request/401/allow-tools-and-deployers-to-specify/diff">this pull request</a>.</p>

<h3 id="prerequisites-enable-docker-to-run-using-sudo-without-a-password">Prerequisites: enable docker to run using sudo without a password</h3>

<p>Add docker runner to sudoers file (replace galaxy with the username you are running galaxy under):</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>galaxy  ALL = (root) NOPASSWD: SETENV: /usr/bin/docker
</code></pre></div></div>

<h3 id="run-galaxy">Run Galaxy</h3>

<ol>
  <li>
    <p>Download Galaxy</p>
  </li>
  <li>
    <p>Edit config/job_conf.xml adding docker runner destination, instructing Galaxy to run dockerized tools.</p>

    <p>Construct a basic job_conf.xml with the following command.</p>
    <div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>cp job_conf.xml.sample_basic job_conf.xml
</code></pre></div>    </div>
    <p>Add a docker destination in job_conf.xml to enable running through docker:</p>
    <div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>&lt;destinations default="docker_local"&gt;
   &lt;destination id="local" runner="local"/&gt;
   &lt;destination id="docker_local" runner="local"&gt;
     &lt;param id="docker_enabled"&gt;true&lt;/param&gt;
   &lt;/destination&gt;
&lt;/destinations&gt;
</code></pre></div>    </div>
    <p>More information can be found in the job_conf.xml.sample_advanced file that comes with Galaxy.</p>
  </li>
  <li>
    <p>BWA is available via ToolShed:
<img src="http://localhost:4000/images/06_bwa_shedtool.png" alt="BWA_shedtool" /></p>

    <p>Setup bwa_wrapper.xml tool. It is located <a href="https://raw.githubusercontent.com/galaxyproject/tools-devteam/master/legacy/bwa_wrappers/bwa_wrapper.xml">here</a></p>

    <div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>mkdir tools/BWA
wget https://raw.githubusercontent.com/galaxyproject/tools-devteam/master/legacy/bwa_wrappers/bwa_wrapper.xml tools/BWA
</code></pre></div>    </div>

    <p>Now, in the tool_conf.xml file please add a new section for this tool:</p>
    <div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>&lt;section&gt;
  &lt;tool file="BWA/bwa_wrapper.xml"/&gt;
 &lt;/section&gt;
</code></pre></div>    </div>

    <p>Edit the xml file, changing</p>
    <div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>&lt;requirements&gt;
  &lt;requirement type="package" version="0.5.9"&gt;bwa&lt;/requirement&gt;
&lt;/requirements&gt;
</code></pre></div>    </div>
    <p>to</p>
    <div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>&lt;requirements&gt;
  &lt;container type="docker"&gt;mtangaro/galaxy-bwa-0.5.9&lt;/container&gt;
&lt;/requirements&gt;
</code></pre></div>    </div>

    <p>Remove interpreter from the command attribute by changing</p>
    <div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>&lt;command interpreter="python"&gt;
  bwa_wrapper.py 
</code></pre></div>    </div>
    <p>to</p>
    <div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>&lt;command&gt;
  bwa_wrapper.py
</code></pre></div>    </div>
    <p>The resulting xml wrapper is available <a href="https://raw.githubusercontent.com/mtangaro/galaxy/Galaxy4Developers/tools/BWA/bwa_wrapper.xml">here</a></p>

    <p>Galaxy is ready!</p>
  </li>
</ol>

<h3 id="bwa-dockerfile">Bwa Dockerfile</h3>

<p>Bwa-galaxy Dockerfile is available <a href="https://github.com/mtangaro/galaxy-bwa-0.5.9">here</a>
Docker Hub image <a href="https://hub.docker.com/r/mtangaro/galaxy-bwa-0.5.9/">here</a></p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>sudo docker pull mtangaro/galaxy-bwa-0.5.9

</code></pre></div></div>
<p>bwa_wrapper.py in the docker container.</p>

<p>Dockerfile:</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>FROM ubuntu:14.04

RUN apt-get -y update
RUN apt-get install -y make build-essential zlib1g-dev python git

### install bwa
ADD https://github.com/lh3/bwa/archive/0.5.9.tar.gz /tmp/bwa.tar.gz
WORKDIR /tmp
RUN tar xvzf /tmp/bwa.tar.gz \
      &amp;&amp; cd /tmp/bwa-0.5.9 \
      &amp;&amp; make \
      &amp;&amp; ln -s /tmp/bwa-0.5.9/bwa /usr/bin/

### get bwa wrapper
RUN mkdir /tmp/bwa
WORKDIR /tmp/bwa
RUN git clone https://github.com/galaxyproject/tools-devteam.git bwa_deps
RUN cp bwa_deps/legacy/bwa_wrappers/bwa_wrapper.py /usr/bin/bwa_wrapper.py
RUN chmod a+x /usr/bin/bwa_wrapper.py
</code></pre></div></div>

<h3 id="hands-on-4">Hands On! (4)</h3>

<p>Remember: Galaxy run docker using sudo. You have to add docker runner to sudoers file.</p>

<ol>
  <li>Everything is already available <a href="https://github.com/mtangaro/galaxy">here</a>.<br />
Clone Galaxy repository fork, which includes all modified files:
    <div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>git clone https://github.com/mtangaro/galaxy.git
cd galaxy
git checkout Galaxy4Developers
./run.sh
</code></pre></div>    </div>
  </li>
  <li>
    <p>Load Input files from here: <a href="https://zenodo.org/record/61377#.WVJFyq0lGJL">link</a></p>

    <p><a href="https://zenodo.org/record/61377/files/hg19_chr8.fa">Reference file</a> - fasta datatype<br />
<a href="https://zenodo.org/record/61377/files/proband_R1.fq">Forward</a> - fastqsanger datatype<br />
<a href="https://zenodo.org/record/61377/files/proband_R2.fq">Reverse </a> - fastqsanger datatype</p>
  </li>
  <li>
    <p>Check BWA docker on your local machine: no BWA dependencies on your machine
<img src="http://localhost:4000/images/06_bwa_no_docker_image.png" alt="06_bwa_no_docker_image" /></p>
  </li>
  <li>
    <p>Configure and Run BWA job:
<img src="http://localhost:4000/images/06_bwa_config.png" alt="06_bwa_config" /></p>
  </li>
  <li>
    <p>Run BWA. Galaxy will automatically download bwa docker image:
<img src="http://localhost:4000/images/06_bwa_docker_image.png" alt="06_bwa_docker_image" /></p>
  </li>
  <li>Check results:
<img src="http://localhost:4000/images/06_bwa_results.png" alt="06_bwa_results" /></li>
</ol>

<h2 id="clean-your-environment">Clean your environment</h2>

<p>Stop and remove all your containers</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ docker rm $(docker ps -a -q) -f
</code></pre></div></div>

<p>Remove all your images</p>
<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ docker rmi $(docker images -q)
</code></pre></div></div>

<h2 id="references">References</h2>

<p>[Reference this tutorial was take from Traning course] (https://crs4.github.io/Galaxy4Developers/)
<a href="https://docs.docker.com/">Docker documentation</a></p>

<p><a href="https://docs.docker.com/engine/reference/run/">Docker run documentation</a></p>

<p><a href="https://docs.docker.com/engine/reference/builder/">Dockerfile</a></p>

<p><a href="https://docs.docker.com/docker-hub/builds/#create-an-automated-build">Docker automated build</a></p>

<p><a href="https://docs.galaxyproject.org/en/latest/admin/special_topics/mulled_containers.html">Galaxy: Containers for tool dependencies</a></p>


</div>

      </div>
    </div>

    <label for="sidebar-checkbox" class="sidebar-toggle"></label>

    <script>
      (function(document) {
        var toggle = document.querySelector('.sidebar-toggle');
        var sidebar = document.querySelector('#sidebar');
        var checkbox = document.querySelector('#sidebar-checkbox');

        document.addEventListener('click', function(e) {
          var target = e.target;

          if(!checkbox.checked ||
             sidebar.contains(target) ||
             (target === checkbox || target === toggle)) return;

          checkbox.checked = false;
        }, false);
      })(document);
    </script>
  <!--<footer>
        <p>This project is maintained by  Maurizio Polano<br/></p>
        <p><small>Hosted on GitHub Pages &mdash; Theme by <a href="https://github.com/orderedlist">orderedlist</a></small></p>
      </footer>
    </div>
    <script src="javascripts/scale.fix.js"></script>

-->
  </body>
</html>
